File: lib/analyze.ts
// Code analysis — uses Groq (free, fast) with Gemini as fallback

export async function analyzeRepo(params: {
  owner: string
  name: string
  description: string
  language: string
  stars: number
  files: string[]
}): Promise<string> {

  const prompt = `You are a senior software engineer helping a student or junior developer understand a codebase.

Repository: ${params.owner}/${params.name}
Description: ${params.description || 'No description'}
Language: ${params.language}
Stars: ${params.stars}
Files: ${params.files.slice(0, 50).join(', ')}

Write a clear, detailed explanation with these sections:

## What This Project Does
3-4 sentences explaining what it is, what problem it solves, and who uses it. Use simple language.

## How It Is Structured
Describe the folder structure. What does each major folder contain?

## Technology Breakdown
List every major technology and its specific role in this project.

## The Most Important Files
List 5-6 files a new developer must read first. For each: what it does and why it matters.

## How to Understand This Codebase (Step by Step)
Numbered steps — in what order should a developer explore this project?

## Common Questions a New Developer Would Ask
Answer 3-4 typical questions about this project.

Be specific. Be clear. Assume the reader is learning.`

  // Try Groq first (free, fast, no credit card)
  const groqKey = process.env.NEXT_PUBLIC_GROQ_API_KEY
  if (groqKey && groqKey !== 'your-groq-api-key-here' && groqKey.length > 10) {
    try {
      const res = await fetch('https://api.groq.com/openai/v1/chat/completions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${groqKey}`,
        },
        body: JSON.stringify({
          model: 'llama-3.3-70b-versatile',
          messages: [{ role: 'user', content: prompt }],
          max_tokens: 2048,
          temperature: 0.3,
        }),
      })
      const d = await res.json()
      if (res.ok && d.choices?.[0]?.message?.content) {
        return d.choices[0].message.content
      }
      if (d.error) throw new Error(d.error.message)
    } catch (e: any) {
      console.warn('Groq failed:', e.message, '— trying Gemini fallback')
    }
  }

  // Gemini fallback
  const geminiKey = process.env.NEXT_PUBLIC_GEMINI_API_KEY
  if (geminiKey && geminiKey !== 'your-gemini-api-key-here' && geminiKey.length > 10) {
    // Try multiple model names in order
    const models = ['gemini-1.5-flash-latest', 'gemini-2.0-flash-exp', 'gemini-1.5-flash']
    for (const model of models) {
      try {
        const res = await fetch(
          `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${geminiKey}`,
          {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }] }),
          }
        )
        const d = await res.json()
        if (res.ok && !d.error) {
          const text = d.candidates?.[0]?.content?.parts?.[0]?.text
          if (text) return text
        }
        if (d.error?.status === 'NOT_FOUND') continue // try next model
        if (d.error) throw new Error(d.error.message)
      } catch (e: any) {
        if (e.message?.includes('NOT_FOUND')) continue
        throw e
      }
    }
  }

  throw new Error('NO_KEY')
}

export async function analyzeFile(params: {
  owner: string
  name: string
  path: string
  content: string
}): Promise<string> {

  const prompt = `Explain this code file to a student or junior developer.

Repository: ${params.owner}/${params.name}
File: ${params.path}

\`\`\`
${params.content.slice(0, 3000)}
\`\`\`

Explain clearly:
## What This File Does
## Key Functions / Classes
## How It Connects to the Rest of the Project

Be specific and simple.`

  // Try Groq first
  const groqKey = process.env.NEXT_PUBLIC_GROQ_API_KEY
  if (groqKey && groqKey !== 'your-groq-api-key-here' && groqKey.length > 10) {
    try {
      const res = await fetch('https://api.groq.com/openai/v1/chat/completions', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${groqKey}` },
        body: JSON.stringify({
          model: 'llama-3.1-8b-instant',
          messages: [{ role: 'user', content: prompt }],
          max_tokens: 1024,
          temperature: 0.2,
        }),
      })
      const d = await res.json()
      if (res.ok && d.choices?.[0]?.message?.content) return d.choices[0].message.content
    } catch {}
  }

  // Gemini fallback
  const geminiKey = process.env.NEXT_PUBLIC_GEMINI_API_KEY
  if (geminiKey && geminiKey !== 'your-gemini-api-key-here' && geminiKey.length > 10) {
    const models = ['gemini-1.5-flash-latest', 'gemini-2.0-flash-exp', 'gemini-1.5-flash']
    for (const model of models) {
      try {
        const res = await fetch(
          `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${geminiKey}`,
          { method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }] }) }
        )
        const d = await res.json()
        if (res.ok && !d.error) {
          const text = d.candidates?.[0]?.content?.parts?.[0]?.text
          if (text) return text
        }
        if (d.error?.status === 'NOT_FOUND') continue
      } catch {}
    }
  }

  throw new Error('NO_KEY')
}

File: backend/routers/analyze.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import httpx, os
from typing import Optional, List

router = APIRouter()

GEMINI_KEY = os.getenv("GEMINI_API_KEY", "")
GEMINI_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent"

async def call_gemini(prompt: str) -> str:
    if not GEMINI_KEY:
        return "⚠️ No GEMINI_API_KEY set. Add it to backend/.env"
    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.post(
            f"{GEMINI_URL}?key={GEMINI_KEY}",
            json={"contents": [{"parts": [{"text": prompt}]}]},
        )
    if res.status_code != 200:
        raise HTTPException(status_code=502, detail="Gemini API error: " + res.text)
    data = res.json()
    return data["candidates"][0]["content"]["parts"][0]["text"]

class RepoAnalyzeRequest(BaseModel):
    owner: str
    name: str
    description: Optional[str] = ""
    language: Optional[str] = ""
    file_paths: List[str] = []

class FileExplainRequest(BaseModel):
    owner: str
    name: str
    path: str
    content: str

@router.post("/repo")
async def analyze_repo(req: RepoAnalyzeRequest):
    prompt = f"""Analyse this GitHub repository and provide a comprehensive technical overview.

Repository: {req.owner}/{req.name}
Description: {req.description or "No description"}
Primary Language: {req.language or "Unknown"}
File tree sample:
{chr(10).join(req.file_paths[:40])}

Provide in markdown:
## What this project does
(2-3 sentence summary)

## Architecture Overview
(Main components and layers)

## Tech Stack
(Each major technology and its role)

## Key Files to Understand First
(Most important files for a new developer)

## Getting Started
(What to read in what order)

Be specific and technical. Use bullet points where helpful."""

    text = await call_gemini(prompt)
    return {"summary": text, "model": "gemini-1.5-flash"}

@router.post("/file")
async def explain_file(req: FileExplainRequest):
    truncated = req.content[:4000] if len(req.content) > 4000 else req.content
    prompt = f"""Explain this code file from the repository {req.owner}/{req.name}.

File path: {req.path}
Content:
```
{truncated}
```

Provide in markdown:
## What this file does
(Concise description)

## Key Functions / Classes
(List the most important ones with brief descriptions)

## How it fits in the project
(Its role in the broader architecture)

Keep it concise, technical, and helpful for a developer new to this codebase."""

    text = await call_gemini(prompt)
    return {"explanation": text, "file": req.path, "model": "gemini-1.5-flash"}

@router.post("/security")
async def security_scan(req: RepoAnalyzeRequest):
    prompt = f"""Perform a security analysis of this repository.

Repository: {req.owner}/{req.name}
Language: {req.language}
Files: {chr(10).join(req.file_paths[:30])}

Identify in markdown:
## Security Risk Summary
(Overall risk level: LOW / MEDIUM / HIGH)

## Potential Vulnerabilities
(List any patterns that could indicate security issues based on file names and structure)

## Recommendations
(Concrete steps to improve security)

Be specific and actionable."""

    text = await call_gemini(prompt)
    return {"report": text, "model": "gemini-1.5-flash"}
